src/main/java/io/debezium/connector/yugabytedb/consistent/Merger.java

```diff
@@ -31,32 +35,51 @@ public void addMessage(Message message) {
         LOGGER.debug("Add message {}", message);
     }
 
-    public Long streamSafeTime() {
+    public BigInteger streamSafeTime() {
         // tabletSafeTime.entrySet().stream().forEach(e -> LOGGER.info("Tablet {}, safe time {}", e.getKey(), e.getValue()));
         return Collections.min(tabletSafeTime.values());
     }
 
+    public long totalQueueSize() {
+        return queue.size();
+    }
+
+    public int pendingMessagesInTablet(String tabletId) {
+        return mergeSlots.get(tabletId).size();
+    }
+
+    public BigInteger safeTimeForTablet(String tabletId) {
+        return tabletSafeTime.get(tabletId);
+    }
+
     public Optional<Message> peek() {
         Message message = queue.peek();
-        Optional<Message> peeked = message != null && message.commitTime < this.streamSafeTime()
+        Optional<Message> peeked = message != null && message.commitTime.compareTo(this.streamSafeTime()) < 0
                 ? Optional.of(message) : Optional.empty();
 
-        if (message != null && message.commitTime < this.streamSafeTime()) {
-            LOGGER.debug("Stream Safe Time {}, Top message is {} ", this.streamSafeTime(), peeked);
+        if (peeked.isPresent() && peeked.get().record.getRowMessage().getOp() == CdcService.RowMessage.Op.INSERT) {
+            LOGGER.debug("Stream Safe Time {}, Top message is {}", this.streamSafeTime(), peeked);
         }
 
         return peeked;
     }
 
-    public Message poll() {
+    public synchronized Optional<Message> poll() {
+        if (queue.isEmpty()) {
+            return Optional.empty();
+        }
+
         Message message = Objects.requireNonNull(queue.poll());
```

Comment:
`poll` removes the message. However, the next if condition maybe false and might lose the message. Instead the code should `peek` instead. 

My suggestion is turn `Merger::peek` to private and then use it in `Merger::poll`. So the code structure is as follows:

```
public synchronised Optional<Message> poll() {
    Message message = this.peek();
    // No valid message to emit
    if (message.isEmpty())
        return message;
        
    // Remove message from queue and mergeSlots
    queue.poll();
    mergeSlots.get(...).removeIf(...);
    return message;    
}
```

Reply:
Addressed.